tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/MetricsHelper.java

```diff
@@ -0,0 +1,236 @@
+   * @return a boolean operand to determine if the shapes have valid dimensions or not.
+   */
+  private static <T extends TNumber> Operand<TBool> hasValidDims(
+      Ops tf, Operand<T> weightsShape, Operand<T> valuesShape) {
+    tf = tf.withSubScope("has_invalid_dims");
+    Operand<T> diff = tf.reduceSum(tf.math.sub(weightsShape, valuesShape), tf.constant(0));
+    return tf.math.equal(cast(tf, tf.constant(0), diff.asOutput().type()), diff);
+  }
+
+  // alias for mean
+
+  /**
+   * Calculate the mean of the operand, along all axes and <code>keepDims</code> is <code>false
+   * </code>
+   *
+   * @param tf the TensorFlow Ops
+   * @param x the Operand used to calculate the mean
+   * @param <T> the type of the Operand.
+   * @return the mean of the operand
+   */
+  public static <T extends TType> Operand<T> mean(Ops tf, Operand<T> x) {
+    return mean(tf, x, null, false);
+  }
+
+  /**
+   * Calculate the mean of the operand, alongside the specified axis with <code>keepDims</code> is
+   * <code>false</code>
+   *
+   * @param tf the TensorFlow Ops
+   * @param x the Operand used to calculate the mean
+   * @param axis Axes to compute the mean.
+   * @param <T> the type of the Operand.
+   * @param <U> the type of the axis.
+   * @return the mean of the operand, alongside the specified axis.
+   */
+  public static <T extends TType, U extends TNumber> Operand<T> mean(
+      Ops tf, Operand<T> x, Operand<U> axis) {
+    return mean(tf, x, axis, false);
+  }
+
+  /**
+   * Calculate the mean of the operand, along all axis.
+   *
+   * @param tf the TensorFlow Ops
+   * @param x the Operand used to calculate the mean
+   * @param keepDims Indicates whether to keep the dimensions or not. If <code>keepdims</code> is
+   *     <code>false</code>, the rank of the tensor is reduced by 1 for each entry in <code>axis
+   *     </code>. If <code>keepdims</code> is <code>true</code>, the reduced dimensions are retained
+   *     with length 1.
+   * @param <T> the type of the operand
+   * @return the mean of elements of <code>x</code>.
+   */
+  public static <T extends TType> Operand<T> mean(Ops tf, Operand<T> x, boolean keepDims) {
+    return mean(tf, x, null, keepDims);
+  }
+
+  /**
+   * Calculate the mean of the operand, alongside the specified axis.
+   *
+   * @param tf the TensorFlow Ops
+   * @param x the Operand used to calculate the mean
+   * @param axis Axes to compute the mean.
+   * @param keepDims Indicates whether to keep the dimensions or not. If `keepdims` is `false`, the
+   *     * rank of the tensor is reduced by 1 for each entry in `axis`. If `keepdims` is `true`, the
+   *     * reduced dimensions are retained with length 1.
+   * @param <T> the data type of the Operand
+   * @param <U> the data type of the axis
+   * @return the mean of elements of `x`.
+   */
+  @SuppressWarnings({"unchecked", "rawtypes"})
+  public static <T extends TType, U extends TNumber> Operand<T> mean(
```

Comment:
The return value of this method won't necessarily have a runtime tensor type of `T`, because `T` could be `TBool`, which could get converted.

Reply:
OK, I added `Z extends TNumber `and return `Operand<Z>`. Also changed `U extends TNumber` to `U extends TIntegral` as U is for shapes.

Reply:
I think the new version still has similar problems. I'm going to thumbs-up this comment as done and add an updated comment to the new version of the code.

👍 moving discussion to new version of code